### ISSUE TITLE
Firenest.PubSub: Is the requirement that topics are always binaries necessary?

### ISSUE THREAD

**Issue by Qqwy** [phoenixframework/firenest#10]
While reading through the source code I remembered something that I have wondered about in the past: What is the reason that the topics to subscribe/publish to are required to be binaries?  If we were to allow tuples of arbitrary BEAM datastructures, wouldn't it mean that the serialization that is done every time a message is broadcast can be avoided?

Or are there some external requirements that mandate the use of binaries for the topic?

**Comment by josevalim** [comment]
Phoenix has reasons I believe but we don't need to carry them over here. We could probably lift them here.

**Comment by josevalim** [comment]
PR please?

**Comment by michalmuskala** [comment]
The problem might be how it will play with other topologies. For example in case of the planned redis topology, we would need to somehow serialize the topic to write it, right? Or am I missing something?

**Comment by Qqwy** [comment]
All right, plane landed. I will make a PR as soon as I'm home, or maybe already in the train if its WiFi works.

@michalmuskala I think we can harness the power of term_to_binary there (which would be something the topology implementation would do - - and only if necessary for that topology's internals - - , rather than the user code havinf to manage this). I think this would even play quite nice if we want to be able to read the stored (e. g. Redis) information from another language, because libraries that understand and parse the Erlang Binary Term format already exist for most of the widespread languages/stacks.

**Comment by Qqwy** [comment]
We do have a choice by the way between:

- lifting the topic restrictions completely
- having a less strict restriction, something like `@type topic :: binary() | atom() | integer() | tuple(topic)` that still doesn't allow odd types like refs, ports, pids, functions. When picking this one we have a choice of only enforcing it in the type, or also enforcing it at runtime with a guard.

Which one should we use and why?

**Comment by josevalim** [comment]
I believe we can lift the topic restrictions because the topic does not
reach the topology. For the topology the topic is part of a message and
that can be any term.

However, there is an open issue to have channels when sending messages, and
in this case, we could use the topic as the channel. On the other hand, I
think it would be best for the topology communication channels to be
integers, and we just hash+partition on top.

In other words, I think lifting the restriction is fine.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by chrismccord** [comment]
@josevalim can we revisit this discussion for a bit? The reason we restricted it is less because we had to, and more because we had no reason to make topics more flexible. As @michalmuskala said, it also made the contract for Phoenix.PubSub simpler when implementing other adapters, but this point may not be relevant when we are talking about it as a regular RPC call. That said, I'd like to hear more usecases around allowing any term before lifting the restriction. For example, monitoring/metrics/reporting is vastly simpler if we assume we have a string topic. Give me a good usecase for non-binary topics and I could be swayed :)

**Comment by josevalim** [comment]
@chrismccord I think we can restrict the topic on Phoenix but should we restrict it at the toolkit level?

**Comment by chrismccord** [comment]
Someone wise once told me it's easy to add things later if people ask for it, but almost impossible to remove things :) 

I don't have a problem lifting the restriction per se, but I don't want us to rush to implement flexibility only because we *can*. For example, I'm imaging folks may want to broadcast to atom topics, because they look nicer, but one thing about topics is they are very often a dynamic thing. Binaries prevent folks from doing `broadcast(:"room:#{id}", ...)` for example.

**Comment by josevalim** [comment]
@chrismccord if we have term topics, we can use `{:room, id}` as the topic, which is more natural than the string?

**Comment by chrismccord** [comment]
Agree, and that looks very nice. I just wanted to hear use cases first before opening up the flexibility. You are right that the phoenix channel integration can sidestep any issues by enforcing binaries so it's less of a concern from that end.

**Comment by chrismccord** [comment]
`{:room, id}` style convinced me for building elixir <-> elixir systems so good to go. Thanks for hearing me out! :)

**Comment by josevalim** [comment]
@chrismccord yeah, i think removing the constraint aligns well with our goals of providing firenest as a more general toolset. I am curious in seeing how the phoenix integration will look like when all is said and done.

**Comment by OvermindDL1** [comment]
Honestly I'm unsure unbounded terms are best, that means if you want to do something like a list of accepted terms as a hierarchical list of topics, and something could subscribe to a list of 2 of them, and something broadcasts to a set of 4 then they still get the message as they get the messages of all the children, would be harder to do (I do something very similar to this currently but I broadcast to each and every level variation, with the levels encoded in the binary with /'s, it's a bit noisy).  I do like tuples and atoms and integers being supported terms though, maybe even refs too...  Not floats unless they have some special handling somehow though, best leave them out...

**Comment by Qqwy** [comment]
@OvermindDL1 so would you like the proposed `@type topic :: binary() | atom() | integer() | tuple(topic)` better?

To be honest, I agree with the "dont add more flexibility than we need" here, because we can always do that later. I do not see a use-case right now where it makes sense for the topic to be anything else than binary, atom, integer, (possibly a reference) or a (nested) tuple of these.

**Comment by josevalim** [comment]
As the Elixir registry, we should support any term.
